### PR TITLE
WIP: Add prometheus_simple receiver

### PR DESCRIPTION
**Pull request by asuresh4**
**Description:** Add `prometheus_simple` receiver. Provides a wrapper around the existing `prometheus` receiver and is intended provide a uniform interface for the `receiver_creator` to initialize receivers.

An example config 

```
receivers:
  prometheus_simple:
    endpoint: "localhost:2379"
    metrics_path: "/metrics"
exporters:
  signalfx:

service:
  pipelines:
    metrics:
      receivers: [prometheus_simple]
      exporters: [signalfx]
```

This is still a WIP.

### CODE REVIEW

File: receiver/simpleprometheusreceiver/receiver.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	pConfig, err := getPrometheusConfig(prw.config)
+
+	if err != nil {
+		prw.logger.Error("Failed to create prometheus receiver config", zap.Error(err))
+	}
+
+	pr, err := pFactory.CreateMetricsReceiver(prw.logger, pConfig, prw.consumer)
+
+	if err != nil {
+		return err
```
Comment (jrcamp) [inline receiver/simpleprometheusreceiver/receiver.go:65]: May want to provide additional error context similar to above.

File: receiver/simpleprometheusreceiver/receiver.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		scrapeConfig,
+	}}
+
+	return out, nil
+}
+
+// Shutdown stops the underlying Prometheus receiver.
+func (prw *prometheusReceiverWrapper) Shutdown(ctx context.Context) error {
+	prw.prometheusRecever.Shutdown(ctx)
+	return nil
```
Comment (jrcamp) [inline receiver/simpleprometheusreceiver/receiver.go:124]:
```suggestion
	return prw.prometheusRecever.Shutdown(ctx)
```

File: receiver/simpleprometheusreceiver/config.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+// Config defines configuration for simple prometheus receiver.
+type Config struct {
+	configmodels.ReceiverSettings `mapstructure:",squash"`
+	// CollectionInterval is the interval at which metrics should be collected
+	CollectionInterval time.Duration `mapstructure:"collection_interval"`
+	// Scheme, http or https.
+	Scheme string `mapstructure:"scheme"`
+	// Target is endpoint to collect metrics from.
+	Target string `mapstructure:"target"`
```
Comment (jrcamp) [inline receiver/simpleprometheusreceiver/config.go:31]: Should be called `endpoint` which ReceiverSettings already has.

File: go.mod [hunk 4]
```diff
@@ -23,13 +23,16 @@ require (
...
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/simpleprometheusreceiver v0.0.0-00010101000000-000000000000
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/wavefrontreceiver v0.0.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/zipkinscribereceiver v0.0.0
 	github.com/pavius/impi v0.0.0-20180302134524-c1cbdcb8df2b
 	github.com/tcnksm/ghr v0.13.0
+	go.uber.org/zap v1.14.1 // indirect
 	golang.org/x/lint v0.0.0-20200130185559-910be7a94367
 	golang.org/x/tools v0.0.0-20200228224639-71482053b885
 	honnef.co/go/tools v0.0.1-2020.1.3
+	k8s.io/client-go v12.0.0+incompatible // indirect
```
[the remark this answers is not in the document]
Comment (asuresh4) [reply go.mod]: Cleaned this up.

File: receiver/simpleprometheusreceiver/receiver.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+// new returns a prometheusReceiverWrapper
+func new(logger *zap.Logger, cfg *Config, consumer consumer.MetricsConsumerOld) *prometheusReceiverWrapper {
+	return &prometheusReceiverWrapper{logger: logger, config: cfg, consumer: consumer}
+}
+
+// Start creates and starts the prometheus receiver.
+func (prw *prometheusReceiverWrapper) Start(ctx context.Context, host component.Host) error {
+	receiverType := (&prometheusreceiver.Factory{}).Type()
+	factory := host.GetFactory(component.KindReceiver, receiverType)
```
[the remark this answers is not in the document]
Comment (asuresh4) [reply receiver/simpleprometheusreceiver/receiver.go]: Yes, that's a good point. Updated to use factory directly.

File: receiver/simpleprometheusreceiver/receiver.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+	pr, err := pFactory.CreateMetricsReceiver(prw.logger, pConfig, prw.consumer)
+	if err != nil {
+		return fmt.Errorf("failed to create prometheus receiver: %v", err)
+	}
+
+	prw.prometheusRecever = pr
+	prw.prometheusRecever.Start(ctx, host)
+
+	return nil
```
Comment (jrcamp) [inline receiver/simpleprometheusreceiver/receiver.go:63]:
```suggestion
	return prw.prometheusRecever.Start(ctx, host)
```

File: receiver/simpleprometheusreceiver/config.go (hunk 3 above)
Comment (asuresh4) [reply receiver/simpleprometheusreceiver/config.go]: Updated to use `configmodels.ReceiverSettings.Endpoint`.

File: receiver/simpleprometheusreceiver/config.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	UseHTTPS          bool   `mapstructure:"use_https"`
+	// Path to the CA cert that has signed the TLS cert.
+	CAFilePath        string `mapstructure:"ca_file_path"`
+	// Path to the client TLS cert to use for TLS required connections.
+	CertFilePath      string `mapstructure:"cert_file_path"`
+	// Path to the client TLS key to use for TLS required connections.
+	KeyFilePath       string `mapstructure:"key_file_path"`
+	// If `use_https` is true and this option is also true, the exporter's TLS
+	// cert will not be verified.
+	SkipVerify        bool   `mapstructure:"skip_verify"`
```
Comment (tigrannajaryan) [inline receiver/simpleprometheusreceiver/config.go:49]:
```suggestion
	InsecureSkipVerify        bool   `mapstructure:"insecure_skip_verify"`
```
It is not clear what "skip_verify" refers to.

File: receiver/simpleprometheusreceiver/config.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+// https://github.com/signalfx/signalfx-agent/blob/master/pkg/core/common/httpclient/http.go.
+type httpOptionsConfig struct {
+	// Whether or not to use HTTPS instead of HTTP.
+	UseHTTPS          bool   `mapstructure:"use_https"`
+	// Path to the CA cert that has signed the TLS cert.
+	CAFilePath        string `mapstructure:"ca_file_path"`
+	// Path to the client TLS cert to use for TLS required connections.
+	CertFilePath      string `mapstructure:"cert_file_path"`
+	// Path to the client TLS key to use for TLS required connections.
+	KeyFilePath       string `mapstructure:"key_file_path"`
```
Comment (tigrannajaryan) [inline receiver/simpleprometheusreceiver/config.go:46]: Do we need "_path" suffixes? What else can a "file" refer to in a configuration? Is "ca_file" ambiguous? Our codebase already uses "cert_file"/"key_file" elsewhere.

Also check `receiver.TLSCredentials` that already exists. It uses "tls_credentials" subkey for TLS settings. Would it make sense to use similar config format here? In that case we will not need "use_https", existence of "tls_credentials" indicates that https is used.

You have 2 additional options here (CAFilePath and SkipVerify) which are only applicable to TLS clients (like Prometheus) so we cannot use `receiver.TLSCredentials` directly, but we can make the rest uniform:

```yaml
receiveers:
  prometheus_simple:
    endpoint: "localhost:1234"
    collection_interval: 30s
    metrics_path: /v2/metrics
    use_service_account: true
    tls_credentials:
      ca_file: "path"
      cert_file: "path"
      key_file: "path"
      insecure_skip_verify: true
```

What do you think?

File: receiver/simpleprometheusreceiver/config.go (hunk 7 above)
Comment (asuresh4) [reply receiver/simpleprometheusreceiver/config.go]: Updated.

File: receiver/simpleprometheusreceiver/config.go (hunk 8 above)
Comment (asuresh4) [reply receiver/simpleprometheusreceiver/config.go]: Yes, I think the `_path` suffix can be dropped. 

Thanks for pointing me to `receiver.TLSCredentials`. I wasn't aware about it. Updated the config keys to match.

Comment (asuresh4) [reply receiver/simpleprometheusreceiver/config.go]: As we add more receivers that use these options we can extend this to handle all such http client config. The SignalFx agent uses [this](https://github.com/signalfx/signalfx-agent/blob/master/pkg/core/common/httpclient/http.go#L13).

**Comment by asuresh4** [comment]
Closing in favor of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/184.

**Pull request closed without merging**